tpm/ak.go

```diff
@@ -0,0 +1,374 @@
+		return fmt.Errorf("failed deleting AK %q: %w", name, err)
+	}
+
+	if err := t.store.DeleteAK(name); err != nil {
+		return fmt.Errorf("failed deleting AK %q from storage: %w", name, err)
+	}
+
+	if err := t.store.Persist(); err != nil {
+		return fmt.Errorf("failed persisting storage: %w", err)
+	}
+
+	return nil
+}
+
+// AttestationParameters returns information about the AK, typically used to
+// generate a credential activation challenge.
+func (ak *AK) AttestationParameters(ctx context.Context) (params attest.AttestationParameters, err error) {
+	if err := ak.tpm.Open(ctx); err != nil {
+		return params, fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer ak.tpm.Close(ctx)
+
+	loadedAK, err := ak.tpm.attestTPM.LoadAK(ak.data)
+	if err != nil {
+		return params, fmt.Errorf("failed loading AK %q: %w", ak.name, err)
+	}
+	defer loadedAK.Close(ak.tpm.attestTPM)
+
+	params = loadedAK.AttestationParameters()
+
+	return
+}
+
+// EncryptedCredential represents encrypted parameters which must be activated
+// against a key.
+type EncryptedCredential attest.EncryptedCredential
+
+// ActivateCredential decrypts the secret using the key to prove that the AK was
+// generated on the same TPM as the EK. This operation is synonymous with
+// TPM2_ActivateCredential.
+func (ak *AK) ActivateCredential(ctx context.Context, in EncryptedCredential) (secret []byte, err error) {
+	if err := ak.tpm.Open(ctx); err != nil {
+		return secret, fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer ak.tpm.Close(ctx)
+
+	loadedAK, err := ak.tpm.attestTPM.LoadAK(ak.data)
+	if err != nil {
+		return secret, fmt.Errorf("failed loading AK %q: %w", ak.name, err)
+	}
+	defer loadedAK.Close(ak.tpm.attestTPM)
+
+	secret, err = loadedAK.ActivateCredential(ak.tpm.attestTPM, attest.EncryptedCredential(in))
+
+	return
+}
+
+// Blobs returns a container for the private and public AK blobs.
+// The resulting blobs are compatible with tpm2-tools, so can be used
+// like this (after having been written to ak.priv and ak.pub):
+//
+//	tpm2_load -C 0x81000001 -u ak.pub -r ak.priv -c ak.ctx
+func (ak *AK) Blobs(ctx context.Context) (*Blobs, error) {
+	if ak.blobs == nil {
+		if err := ak.tpm.Open(ctx); err != nil {
+			return nil, fmt.Errorf("failed opening TPM: %w", err)
+		}
+		defer ak.tpm.Close(ctx)
+
+		aak, err := ak.tpm.attestTPM.LoadAK(ak.data)
+		if err != nil {
+			return nil, fmt.Errorf("failed loading AK: %w", err)
+		}
+		defer aak.Close(ak.tpm.attestTPM)
+
+		public, private, err := aak.Blobs()
+		if err != nil {
+			return nil, fmt.Errorf("failed getting AK blobs: %w", err)
+		}
+		ak.setBlobs(private, public)
+	}
+
+	return ak.blobs, nil
+}
+
+func (ak *AK) SetCertificateChain(ctx context.Context, chain []*x509.Certificate) error {
+	if err := ak.tpm.Open(ctx); err != nil {
+		return fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer ak.tpm.Close(ctx)
+
+	// TODO(hs): perform validation, such as check if the chain includes leaf for the
+	// AK public key?
```

Comment:
Done in 1d8a8b434257ae7d1338ba30b54f9bf8abb93ca4. 

I'm not entirely happy with the approach yet. I think I'll (de)serialize the AK public key, so that it can be more easily reloaded instead of having to interact with the TPM for this.